File: src/main/java/ru/serdtsev/homemoney/moneyoper/model/MoneyOper.java
package ru.serdtsev.homemoney.moneyoper.model;

import org.apache.logging.log4j.util.Strings;
import ru.serdtsev.homemoney.account.Account;
import ru.serdtsev.homemoney.account.AccountRepository;
import ru.serdtsev.homemoney.account.AccountType;
import ru.serdtsev.homemoney.account.Balance;
import ru.serdtsev.homemoney.balancesheet.BalanceSheet;
import ru.serdtsev.homemoney.utils.Utils;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.persistence.*;
import javax.validation.constraints.AssertTrue;
import java.io.Serializable;
import java.math.BigDecimal;
import java.sql.Timestamp;
import java.time.Instant;
import java.time.LocalDate;
import java.util.*;

import static ru.serdtsev.homemoney.moneyoper.model.MoneyOperStatus.*;
import static ru.serdtsev.homemoney.utils.Utils.assertNonNulls;

@Entity
@Table(name = "money_oper")
public class MoneyOper implements Serializable {
  @Id
  private UUID id;

  @ManyToOne
  @JoinColumn(name = "balance_sheet_id")
  @Nonnull
  private BalanceSheet balanceSheet;

  @Enumerated(EnumType.STRING)
  @Nonnull
  private MoneyOperStatus status;

  @Column(name = "trn_date")
  @Nonnull
  private LocalDate performed;

  @Column(name = "date_num")
  private Integer dateNum;

  @OneToMany(cascade = CascadeType.ALL)
  @JoinColumn(name = "oper_id")
  @Nonnull
  private List<MoneyOperItem> items;

  @ManyToMany
  @JoinTable(name = "labels2objs",
      joinColumns = @JoinColumn(name = "obj_id", foreignKey = @ForeignKey(value = ConstraintMode.NO_CONSTRAINT)),
      inverseJoinColumns = @JoinColumn(name = "label_id"))
  private Set<Label> labels;

  private String comment;

  @Enumerated(EnumType.STRING)
  private Period period;

  @Column(name = "created_ts")
  @Nonnull
  private Timestamp created;

  @OneToOne
  @JoinColumn(name = "parent_id")
  private MoneyOper parentOper;

  @Column(name = "from_acc_id")
  private UUID fromAccId;

  @Column(name = "to_acc_id")
  private UUID toAccId;

  @Column(name = "amount")
  private BigDecimal amount;

  @Column(name = "to_amount")
  private BigDecimal toAmount;

  /**
   * Идентификатор повторяющейся операции. Служит для получения списка операций, которые были созданы по одному шаблону.
   */
  @Column(name = "recurrence_id")
  private UUID recurrenceId;

  public MoneyOper() {
  }

  public MoneyOper(UUID id, BalanceSheet balanceSheet, MoneyOperStatus status, LocalDate performed, Integer dateNum,
      Collection<Label> labels, String comment, Period period) {
    this.id = id;
    this.balanceSheet = balanceSheet;
    this.status = status;
    this.performed = performed;
    this.dateNum = dateNum;
    this.items = new ArrayList<>();
    this.labels = new HashSet<>(labels);
    this.comment = comment;
    this.period = period;
    this.created = Timestamp.from(Instant.now());
  }

  public UUID getId() {
    return id;
  }

  public BalanceSheet getBalanceSheet() {
    return balanceSheet;
  }

  private void setStatus(MoneyOperStatus status) {
    this.status = status;
  }

  public MoneyOperStatus getStatus() {
    return status;
  }

  public LocalDate getPerformed() {
    return performed;
  }

  public void setPerformed(LocalDate performed) {
    this.performed = performed;
  }

  public List<MoneyOperItem> getItems() {
    return items;
  }

  public void setItems(List<MoneyOperItem> items) {
    this.items = items;
  }

  public Period getPeriod() {
    return period;
  }

  public void setPeriod(Period period) {
    this.period = period;
  }

  public String getComment() {
    return Utils.nvl(comment, Strings.EMPTY);
  }

  public void setComment(String comment) {
    this.comment = comment;
  }

  public Collection<Label> getLabels() {
    return Collections.unmodifiableSet(labels);
  }

  public void setLabels(Collection<Label> labels) {
    this.labels.retainAll(labels);
    this.labels.addAll(labels);
  }

  public Integer getDateNum() {
    return dateNum;
  }

  public void setDateNum(Integer dateNum) {
    this.dateNum = dateNum;
  }

  public void setParentOper(MoneyOper parentOper) {
    this.parentOper = parentOper;
  }

  public MoneyOper getParentOper() {
    return parentOper;
  }

  public UUID getParentOperId() {
    return parentOper != null ? parentOper.getId() : null;
  }

  public Timestamp getCreated() {
    return created;
  }

  public void setFromAccId(UUID fromAccId) {
    this.fromAccId = fromAccId;
  }

  @Deprecated
  public UUID getFromAccId() {
    return fromAccId;
  }

  public void setToAccId(UUID toAccId) {
    this.toAccId = toAccId;
  }

  @Deprecated
  public UUID getToAccId() {
    return toAccId;
  }

  public void setAmount(BigDecimal amount) {
    this.amount = amount;
  }

  @Deprecated
  public BigDecimal getAmount() {
    return items.stream()
        .map(MoneyOperItem::getValue)
        .filter(value -> value.signum() < 0)
        .map(BigDecimal::abs)
        .findFirst()
        .orElse(amount);
  }

  @Deprecated
  public String getCurrencyCode() {
    return items.stream()
        .sorted(Comparator.comparingInt(item -> item.getValue().signum()))
        .map(item -> item.getBalance().getCurrencyCode())
        .findFirst()
        .orElse(getBalanceSheet().getCurrencyCode());
  }

  public void setToAmount(BigDecimal toAmount) {
    this.toAmount = toAmount;
  }

  @Deprecated
  public BigDecimal getToAmount() {
    return items.stream()
        .map(MoneyOperItem::getValue)
        .filter(value -> value.signum() > 0)
        .map(BigDecimal::abs)
        .findFirst()
        .orElse(toAmount);
  }

  public String getToCurrencyCode() {
    return items.stream()
        .sorted(Comparator.comparingInt(item -> item.getValue().signum() * -1))
        .map(item -> item.getBalance().getCurrencyCode())
        .findFirst()
        .orElse(getBalanceSheet().getCurrencyCode());
  }

  public UUID getRecurrenceId() {
    return recurrenceId;
  }

  public void setRecurrenceId(UUID recurrenceId) {
    this.recurrenceId = recurrenceId;
  }

  public MoneyOperType getType() {
    if (items.stream().findFirst().orElseThrow(IllegalStateException::new).getBalance().getType() == AccountType.reserve) {
      return MoneyOperType.transfer;
    }
    Optional<Integer> valueSignumSumOpt = items.stream()
        .map(item -> item.getValue().signum())
        .reduce((a, s) -> a += s);
    int valueSignumSum = valueSignumSumOpt.orElseThrow(() ->
        new IllegalStateException("items is empty: " + id));
    if (valueSignumSum > 0) {
      return MoneyOperType.income;
    } else if (valueSignumSum < 0) {
      return MoneyOperType.expense;
    }
    return MoneyOperType.transfer;
  }

  @Deprecated
  public MoneyOperType getType(AccountRepository accountRepo) {
    Account fromAcc = accountRepo.findOne(fromAccId);
    Account toAcc = accountRepo.findOne(toAccId);
    if (fromAcc.getType().equals(AccountType.income)) {
      return MoneyOperType.income;
    }
    if (toAcc.getType().equals(AccountType.expense)) {
      return MoneyOperType.expense;
    }
    return MoneyOperType.transfer;
  }

  public void complete() {
    assert getStatus() == pending || getStatus() == cancelled : getStatus();
    assert !performed.isAfter(LocalDate.now());
    changeBalances(false);
    setStatus(done);
  }

  public void cancel() {
    assert getStatus() == done || getStatus() == pending || getStatus() == template;
    if (getStatus() == done) {
      changeBalances(true);
    }
    setStatus(cancelled);
  }

  public void addItems(Collection<MoneyOperItem> items) {
    items.forEach(item -> addItem(item.getBalance(), item.getValue()));
  }

  public MoneyOperItem addItem(Balance balance, BigDecimal value) {
    return addItem(balance, value, LocalDate.now());
  }

  public MoneyOperItem addItem(Balance balance, BigDecimal value, @Nullable LocalDate performed) {
    assertNonNulls(balance, value);
    assert value.compareTo(BigDecimal.ZERO) != 0 : this.toString();
    MoneyOperItem item = new MoneyOperItem(UUID.randomUUID(), this, balance, value, performed, items.size());
    items.add(item);
    if (value.signum() ==  -1) {
      fromAccId = balance.getId();
      amount = value.abs();
    } else if (value.signum() ==  1) {
      toAccId = balance.getId();
      toAmount = value;
    }
    return item;
  }

  public void changeBalances(boolean revert) {
    BigDecimal factor = revert ? BigDecimal.ONE.negate() : BigDecimal.ONE;
    items.forEach(item -> item.getBalance().changeValue(item.getValue().multiply(factor), this));
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;
    MoneyOper moneyOper = (MoneyOper) o;
    return Objects.equals(id, moneyOper.id);
  }

  @Override
  public int hashCode() {
    return Objects.hash(id);
  }

  public boolean essentialEquals(MoneyOper other) {
    assert this.equals(other);
    return itemsEssentialEquals(other);
  }

  boolean itemsEssentialEquals(MoneyOper other) {
    return items.stream().allMatch(item ->
        other.getItems().stream().anyMatch(i -> i.equals(item)));
  }

  @Override
  public String toString() {
    return "MoneyOper{" +
        "id=" + id +
        ", balanceSheet=" + balanceSheet +
        ", status=" + status +
        ", performed=" + performed +
        ", items=" + items +
        ", created=" + created +
        '}';
  }

  @AssertTrue(message = "Fields amount and toAmount is different.")
  public boolean isAmountsValid() {
    return !Objects.equals(getCurrencyCode(), getToCurrencyCode()) || amount.compareTo(toAmount) == 0;
  }

  @AssertTrue(message = "Field recurrenceId of template is null.")
  public boolean isRecurrenceIdNotNullForTemplate() {
    return status != MoneyOperStatus.template || Objects.nonNull(recurrenceId);
  }
}

File: src/main/kotlin/ru/serdtsev/homemoney/moneyoper/model/MoneyOper.kt
package ru.serdtsev.homemoney.moneyoper.model

import ru.serdtsev.homemoney.account.model.AccountType
import ru.serdtsev.homemoney.account.model.Balance
import ru.serdtsev.homemoney.balancesheet.model.BalanceSheet
import ru.serdtsev.homemoney.common.Model
import java.io.Serializable
import java.math.BigDecimal
import java.sql.Timestamp
import java.time.Instant
import java.time.LocalDate
import java.time.temporal.ChronoUnit
import java.util.*

class MoneyOper (
    val id: UUID,
    val balanceSheet: BalanceSheet,
    var items: MutableList<MoneyOperItem> = mutableListOf(),
    var status: MoneyOperStatus,
    var performed: LocalDate = LocalDate.now(),
    var dateNum: Int? = 0,
    tags: Collection<Tag> = mutableListOf(),
    comment: String? = null,
    var period: Period? = null
) : Model, Serializable {
    var created: Timestamp = Timestamp.from(Instant.now().truncatedTo(ChronoUnit.MILLIS))
    val tags: MutableSet<Tag> = tags.toMutableSet()
    var parentOper: MoneyOper? = null
    /**
     * Идентификатор повторяющейся операции. Служит для получения списка операций, которые были созданы по одному шаблону.
     */
    var recurrenceId: UUID? = null

    var comment: String? = comment
        get() = field.orEmpty()

    val type: MoneyOperType
        get() {
            val hasReserve = items.any { it.balance.type == AccountType.reserve }
            if (!hasReserve) {
                val valueSignedSum = items.map { it.value.signum() }.sum()
                if (valueSignedSum > 0) {
                    return MoneyOperType.income
                } else if (valueSignedSum < 0) {
                    return MoneyOperType.expense
                }
            }
            return MoneyOperType.transfer
        }

    constructor(balanceSheet: BalanceSheet, status: MoneyOperStatus, performed: LocalDate? = LocalDate.now(),
        dateNum: Int? = 0, tags: Collection<Tag>? = mutableListOf(), comment: String? = null, period: Period? = Period.month
    ) : this(UUID.randomUUID(), balanceSheet, mutableListOf(), status, performed!!, dateNum, tags!!, comment, period)

    fun addItem(balance: Balance, value: BigDecimal, performed: LocalDate = this.performed,
            index: Int = items.size, id: UUID = UUID.randomUUID()): MoneyOperItem {
        val item = MoneyOperItem(id, this.id, balance, value, performed, index)
        items.add(item)
        return item
    }

    fun setTags(tags: Collection<Tag>) {
        this.tags.retainAll(tags.toSet())
        this.tags.addAll(tags)
    }

    fun getParentOperId(): UUID? = parentOper?.id

    val isForeignCurrencyTransaction: Boolean
        get() = items.any { it.balance.currencyCode != balanceSheet.currencyCode }

    val valueInNationalCurrency: BigDecimal
        get() = items
                .filter { it.balance.currencyCode == balanceSheet.currencyCode }
                .map { it.value }
                .reduce { acc, value -> acc.add(value) }

    fun complete() {
        assert(status == MoneyOperStatus.pending || status == MoneyOperStatus.cancelled) { status }
        assert(!performed.isAfter(LocalDate.now()))
        changeBalances(false)
        status = MoneyOperStatus.done
    }

    fun cancel() {
        assert(status == MoneyOperStatus.done || status == MoneyOperStatus.pending || status == MoneyOperStatus.template)
        if (status == MoneyOperStatus.done) {
            changeBalances(true)
        }
        status = MoneyOperStatus.cancelled
    }

    fun getBalances(): List<Balance> {
        return items.map { it.balance }
    }

    fun changeBalances(revert: Boolean) {
        val factor = BigDecimal.ONE.let { if (revert) it.negate() else it }
        items.forEach { item -> item.balance.changeValue(item.value * factor, this) }
    }

    override fun toString(): String {
        return "MoneyOper{" +
                "id=" + id +
                ", balanceSheet=" + balanceSheet +
                ", status=" + status +
                ", performed=" + performed +
                ", items=" + items +
                ", created=" + created +
                '}'
    }

    override fun merge(other: Any): Collection<Model> = merge(other as MoneyOper, this)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as MoneyOper

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    companion object {
        fun merge(from: MoneyOper, to: MoneyOper): Collection<Model> {
            assert(from.id == to.id)
            val changedModels = mutableSetOf<Model>(to)
            val prevStatus = from.status
            val mostlyEquals = balanceEquals(from, to)
            if (!mostlyEquals && to.status == MoneyOperStatus.done) {
                to.cancel()
            }
            changedModels.addAll(to.items.map { it.balance })

            mergeItems(from, to)
            to.performed = from.performed
            to.setTags(from.tags)
            to.dateNum = from.dateNum
            to.period = from.period
            to.comment = from.comment

            if (!mostlyEquals && prevStatus == MoneyOperStatus.done
                || to.status == MoneyOperStatus.pending && from.status == MoneyOperStatus.done
            ) {
                to.complete()
            }

            changedModels.addAll(to.items.map { it.balance })
            return changedModels;
        }

        private fun mergeItems(from: MoneyOper, to: MoneyOper) {
            to.items.forEach { item -> from.items.firstOrNull { it == item }?.let { MoneyOperItem.merge(it, item) } }
            to.items.removeIf { item -> from.items.none { it.id == item.id } }
            to.items.addAll(from.items.filter { item -> to.items.none { it.id == item.id } })
        }

        /**
         * Возвращает true, если экземпляры одной операции эквивалентны по проводкам
         */
        fun balanceEquals(a: MoneyOper, b: MoneyOper): Boolean {
            assert(b.id == a.id)
            return b.type == a.type
                    && a.items.all { item -> b.items.any { it.id == item.id && MoneyOperItem.balanceEquals(it, item) }
            }
        }
    }

}